tests/store/index-level.spec.ts

```diff
@@ -1,136 +1,191 @@
-        'c'         : 'd'
-      });
+    it('fails to index with no sorting properties', async () => {
+      const id = uuid();
 
-      const keys = await ArrayUtility.fromAsyncGenerator(testPartition.keys());
-      expect(keys.length).to.equal(4);
+      let failedIndex = testIndex.put(tenant, id, id, {
+        foo: 'foo'
+      }, { nested: {} });
+
+      await expect(failedIndex).to.eventually.be.rejectedWith('must include at least one sorted index');
+
+      failedIndex = testIndex.put(tenant, id, id, {
+        foo: 'foo'
+      }, { sort: [ [] ] });
+
+      await expect(failedIndex).to.eventually.be.rejectedWith('must include at least one sorted index');
+
+      const keys = await ArrayUtility.fromAsyncGenerator(partitionedDB.keys());
+      expect(keys.length).to.equal(0);
+
+      failedIndex = testIndex.put(tenant, id, id, {
+        foo: 'foo'
+      }, { sort: id });
+      await expect(failedIndex).to.eventually.not.be.rejected;
     });
 
-    it('flattens nested records', async () => {
+    it('fails to index with no indexable properties ', async () => {
       const id = uuid();
-      const doc = {
+
+      let failedIndex = testIndex.put(tenant, id, id, {}, { id });
+      await expect(failedIndex).to.eventually.be.rejectedWith('must include at least one indexable property');
+
+      failedIndex = testIndex.put(tenant, id, id, {
+        empty: [ [] ]
+      }, { id });
+      await expect(failedIndex).to.eventually.be.rejectedWith('must include at least one indexable property');
+
+      failedIndex = testIndex.put(tenant, id, id, {
+        foo : {},
+        bar : {
+          baz: {},
+        }
+      }, { id });
+
+      await expect(failedIndex).to.eventually.be.rejectedWith('must include at least one indexable property');
+
+      const keys = await ArrayUtility.fromAsyncGenerator(partitionedDB.keys());
+      expect(keys.length).to.equal(0);
+
+      failedIndex = testIndex.put(tenant, id, id, {
+        foo : 'foo',
+        bar : {
+          baz: 'baz'
+        }
+      }, { id });
+      await expect(failedIndex).to.eventually.not.be.rejected;
+    });
+
+    it('flattens nested indexes', async () => {
+      const id = uuid();
+      const index = {
         some: {
           nested: {
             object: true
           }
         }
       };
-      await index.put(tenant, id, doc);
-
-      const key = await testPartition.get(index['join']('some.nested.object', true, id));
-      expect(key).to.equal(id);
+      await testIndex.put(tenant, id, id, index, { id });
+      const indexKey = testIndex['constructIndexedKey'](
+        `id`,
+        'some.nested.object',
+        'true',
+        `"${id}"`,
+        id,
+      );
+      const key = await partitionedDB.get(indexKey);
+      expect(key).to.not.be.undefined;
+      expect(JSON.parse(key!)).to.equal(id);
     });
 
-    it('removes empty objects', async () => {
+    it('adds 1 key per property, per sorted property, aside from id', async () => {
```

Comment:
This test made me realize that I don't know the expected behavior of `put`ing the same key-value pair twice. I would have thought `put`ting to the same key without deleting would either:
1. Overwrite the existing value and indexes.
2. Throw error

But it looks like we actually have a third behavior:
3. Value is overwritten, keep existing indexes, AND add new indexes.

@thehenrytsai Curious your thoughts here. What do we expect to happen to existing indexes if we `IndexLeve#put` to the same key twice without deletion? How about if we `MessageStore#put` to the same `messageCid` twice? Or should it never happen that we `put` to the same key twice?

Reply:
Alternate implementations of the `MessageStore` and `EventLog` should have the same behavior as our `level` implementation`. I haven't yet investigated how `dwn-sql-store` handles double-`put`s, but it's probably subtly different than `IndexLevel`.

Reply:
This is a good call out.

It *technically* shouldn't arrive at indexing the same record twice, as it should be met with a 409 error somewhere before it arrives here.

However maybe it's worth putting in some defensive programming in place for this and explicitly ignore.

Looking at it there is definitely a different in behavior between MessageStore and EventLog with the same data, and there is also a difference in behavior between the dwn-sql-store implementation (which just uses an incremented ID as a unique index in both cases).

@thehenrytsai mentioned that we might be able to just combine MessageStore and EventLog, which I want to think about today and see what the implications would be.

It makes sense in the dwn-sql-store as they are literally indexing the same things, the query is just slightly different.